core/types/rollup_cost.go

```diff
@@ -116,9 +185,50 @@ func newL1CostFunc(l1Basefee, overhead, scalar *big.Int, isRegolith bool) l1Cost
 	}
 }
 
+// newL1CostFuncEcotone returns an l1 cost function suitable for the Ecotone upgrade except for the
+// very first block of the upgrade.
+func newL1CostFuncEcotone(l1Basefee, l1BlobBasefee, l1BasefeeScalar, l1BlobBasefeeScalar *big.Int) l1CostFunc {
+	return func(costData RollupCostData) (fee, calldataGasUsed *big.Int) {
+		calldataGas := (costData.zeroes * params.TxDataZeroGas) + (costData.ones * params.TxDataNonZeroGasEIP2028)
+		calldataGasUsed = new(big.Int).SetUint64(calldataGas)
+
+		// Ecotone L1 cost function:
+		//
+		//   (gas/16)*(l1Basefee*16*l1BasefeeScalar + l1BlobBasefee*l1BlobBasefeeScalar)/1e6
+		//
+		// We divide "gas" by 16 to change from units of calldata gas to "estimated # of bytes when
+		// compressed".
+		//
+		// Function is actually computed as follows for better precision under integer arithmetic:
+		//
+		//   gas*(l1Basefee*16*l1BasefeeScalar + l1BlobBasefee*l1BlobBasefeeScalar)/16e6
```

Comment:
```suggestion
		//   (calldataGas/16)*(l1Basefee*16*l1BasefeeScalar + l1BlobBasefee*l1BlobBasefeeScalar)/1e6
		//
		// We divide "calldataGas" by 16 to change from units of calldata gas to "estimated # of bytes when
		// compressed". Known as "compressedTxSize" in the spec.
		//
		// Function is actually computed as follows for better precision under integer arithmetic:
		//
		//   calldataGas*(l1Basefee*16*l1BasefeeScalar + l1BlobBasefee*l1BlobBasefeeScalar)/16e6
```

Naming it `calldataGas` instead of `gas` to be less confusing, and so we can use the same thing in the spec. Also referencing `compressedTxSize` of the spec.

Doing the higher-precision preserving calculation is fine, but it's not defined in the spec. I think we need to update the specs to be precise about the order of operations for the precision to be preserved, and consistent across different implementations.

Reply:
It is defined in the spec, however probably not clearly enough:

```
 (To preserve precision under integer arithmetic, the division by 16 is actually
  performed at the very end of the fee computation together with the division by 1e6 as a single
  division by 16e6.)
```